### PR TITLE
Fix tmp filename generation

### PR DESCRIPTION
**Pull request by banasraf**
Autoserialize was not working correctly due to race condition in tmp file naming. It was using a timestamp with a resolution of a second which caused conflicts between models.

To fix this, I changed the timestamp to maximal resolution of a monotonic clock (std::steady_clock).

### CODE REVIEW

File: src/utils/utils.h [hunk 1]
```diff
@@ -88,11 +88,8 @@ std::string vec_to_string(const std::vector<T> &vec, const std::string &lbracket
 
 
 inline std::string timestamp() {
-  std::time_t t = std::time(nullptr);
-  std::tm tm = *std::localtime(&t);
-  std::stringstream timestamp;
-  timestamp << std::put_time(&tm, "%Y%m%d_%H%M%S");
-  return timestamp.str();
+  auto timestamp = std::chrono::steady_clock::now();
+  return std::to_string(timestamp.time_since_epoch().count());
```
[the remark this answers is not in the document]
Comment (mzient) [reply src/utils/utils.h:92]: If you want a unique timestamp, this should do the trick (I think):
```suggestion
  auto timestamp = std::chrono::steady_clock::now();
  int64_t ts = timestamp.time_since_epoch().count();
  static std::atomic<int64_t> prev_ts = 0;
  int64_t expected = prev_ts.load();
  while (!prev_ts.compare_exchange_strong(expected, ts)) {
    expected = prev_ts.load();  // this should not be necessary but I remember reading about some bug in libc++
    ts = std::max(timestamp.time_since_epoch().count(), expected + 1);
  }
  return std::to_string(ts);
```
You can also not call it "timestamp" but rather "tmpname" and use a simpler approach:
```C++
  static std::atomic<int64_t> ts(std::chrono::steady_clock::now().time_since_epoch().count);
  return ts++;
```

[the remark this answers is not in the document]
Comment (banasraf) [reply src/utils/utils.h]: I've decided to go with the simplest solution with the atomic.

[the remark this answers is not in the document]
Comment (JanuszL) [reply src/utils/utils.h]: According to https://stackoverflow.com/questions/32463299/get-the-file-name-that-was-created-by-mkstemp it should be enough to (with the guarantee of uniqueness):
```
inline std::string tmpname() {
  char template[] = "/tmp/fileXXXXXX";
  close(mkstemp(template));
  return template;
```

[the remark this answers is not in the document]
Comment (JanuszL) [reply src/utils/utils.h]: But up to you

[the remark this answers is not in the document]
Comment (banasraf) [reply src/utils/utils.h]: Ok, decided to go with mkstemp, because it's completely bulletproof.

**Review by mzient — APPROVED** [review]

File: src/utils/utils.h [hunk 2]
```diff
@@ -23,11 +23,12 @@
 #ifndef DALI_BACKEND_UTILS_UTILS_H_
 #define DALI_BACKEND_UTILS_UTILS_H_
 
-#include <ctime>
+#include <chrono>
 #include <string>
 #include <sstream>
 #include <vector>
 #include <iomanip>
+#include <atomic>
```
Comment (JanuszL) [inline src/utils/utils.h:31]: Not needed anymore.

**Review by JanuszL — DISMISSED** [review]

**Review by JanuszL — APPROVED** [review]

**Review by szalpal — COMMENTED** [review]
Forgot to approve before merging 😅  This is an approval.

**Pull request merged**